clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java

```diff
@@ -138,72 +141,127 @@ public TaskHealthCachingAgent(
+
       DescribeTargetHealthResult describeTargetHealthResult =
           amazonloadBalancing.describeTargetHealth(
               new DescribeTargetHealthRequest()
                   .withTargetGroupArn(loadBalancer.getTargetGroupArn())
                   .withTargets(
                       new TargetDescription()
                           .withId(networkInterface.getPrivateIpv4Address())
-                          .withPort(
-                              taskDefinition
-                                  .getContainerDefinitions()
-                                  .get(0)
-                                  .getPortMappings()
-                                  .get(0)
-                                  .getContainerPort())));
+                          .withPort(targetGroupPort.get())));
 
       if (describeTargetHealthResult.getTargetHealthDescriptions().isEmpty()) {
+        log.debug("Target health description is empty");
         evictStaleData(task, loadBalancerService);
         continue;
       }
 
+      log.debug(
+          "Target health description is not empty and has a size of {}",
+          describeTargetHealthResult.getTargetHealthDescriptions().size());
       TargetHealthDescription healthDescription =
           describeTargetHealthResult.getTargetHealthDescriptions().get(0);
 
       TaskHealth taskHealth = makeTaskHealth(task, serviceName, healthDescription);
-      return taskHealth;
+      taskHealthList.add(taskHealth);
+    }
+    return taskHealthList;
+  }
+
+  private Optional<Integer> getTargetGroupContainerPort(
+      List<ContainerDefinition> containerDefinitions, Integer containerPort) {
+    for (ContainerDefinition containerDefinition : containerDefinitions) {
+      log.debug(
+          "Looking for ContainerPort: {} in PortMappings: {}",
+          containerPort,
+          containerDefinition.getPortMappings());
+      for (PortMapping portMapping : containerDefinition.getPortMappings()) {
+        if (portMapping.getContainerPort().intValue() == containerPort.intValue()) {
+          log.debug("Load balanced containerPort: {} found for container.", containerPort);
+          return Optional.of(containerPort);
+        }
+      }
+    }
+
+    return Optional.empty();
+  }
+
+  private Optional<Integer> getTargetGroupHostPort(
+      List<ContainerDefinition> containerDefinitions, Integer lbHostPort) {
+    for (ContainerDefinition containerDefinition : containerDefinitions) {
+      log.debug(
+          "Looking for HostPort: {} in PortMappings: {}",
+          lbHostPort,
+          containerDefinition.getPortMappings());
+      for (PortMapping portMapping : containerDefinition.getPortMappings()) {
+        if (portMapping.getHostPort().intValue() == lbHostPort.intValue()) {
+          log.debug("Load balanced hostPort: {} found for container.", lbHostPort);
+          return Optional.of(lbHostPort);
+        }
+      }
```

Comment:
How does this work when using dynamic port mapping?  The host port can be set to 0 in the task definition port mapping to let ECS choose a port on the host, which is why previously the code looked at the network bindings for the actual task to find the port

Reply:
This method is only invoked in the inferNetworkBindings method: https://github.com/spinnaker/clouddriver/pull/4218/files/6541bb347f65b37dbfe70488a5f56f8351b234ce#diff-e47e47d8736b959dc299541a637e9327R326-R327

Reply:
Right, network-binded containers are the ones that will have dynamic port mappings.  Network-interfaced containers (i.e. Fargate) will always have LB port == container port, but network-binded containers can have an ECS-assigned LB port.

Reply:
Addressed. Iterating over the network bindings and comparing the host port to the lb container port instead of the container def. Verified that the container port will always map to one of the network binding host ports